### PR TITLE
Remove sensitive config values from diagnostics bundles and build output

### PR DESCRIPTION
**Pull request by branden**
## High Level Description

This adds support for tagging gen config variables as "secret", which removes them from diagnostics bundles and build output.

Secret variables are omitted from `/opt/mesosphere/etc/expanded.config.json`, and their values are masked in `/opt/mesosphere/etc/user.config.yaml` and when they're printed to the console. New files `/opt/mesosphere/etc/user.config.full.yaml` and `/opt/mesosphere/etc/expanded.config.full.json` are added which include secret variables, and these files are only readable by root. 

This also tags some existing config variables as secret:
* `cluster_docker_credentials`
* `exhibitor_admin_password`
* `aws_template_storage_access_key_id`
* `aws_template_storage_secret_access_key`

## Related Issues

  - [DCOS_OSS-1795](https://jira.mesosphere.com/browse/DCOS_OSS-1795) Sensitive config values in diagnostics bundles

## Checklist for all PR's

  - [x] Included a test which will fail if code is reverted but test is not. If there is no test please explain here:
  - [x] Read the [DC/OS contributing guidelines](https://github.com/dcos/dcos/blob/master/contributing.md)
  - [x] Followed relevant code rules [Rules for Packages and Systemd](https://github.com/dcos/dcos/tree/master/docs)

## Checklist for component/package updates:

If you are changing components or packages in DC/OS (e.g. you are bumping the sha or ref of anything underneath `packages`), then in addition to the above please also include:

  - [ ] Change log from the last version integrated (this should be a link to commits for easy verification and review): [example](https://github.com/dcos/dcos-mesos-modules/compare/f6fa27d7c40f4207ba3bb2274e2cfe79b62a395a...6660b90fbbf69a15ef46d0184e36755881d6a5ae)
  - [ ] Test Results: [link to CI job test results for component]
  - [ ] Code Coverage (if available): [link to code coverage report]

### CODE REVIEW

**Comment by branden** [comment]
@mesosphere-mergebot bump-ee

**Comment by branden** [comment]
@mesosphere-mergebot label ready-for-review

File: gen/__init__.py [hunk 1]
```diff
@@ -526,10 +530,25 @@ def get_late_variables(resolver, sources):
     return late_variables
 
 
+def get_secret_variables(sources):
+    return list(set().union(*(source.secret for source in sources)))
```
[the remark this answers is not in the document]
Comment (branden) [reply gen/__init__.py]: This is what happens when I stare at a generator expression for too long

File: gen/internals.py [hunk 2]
```diff
@@ -246,44 +246,62 @@ def __eq__(self, other):
...
-        self.setters.setdefault(name, list()).append(Setter(name, value, is_optional, conditions, self.is_user))
+        self.setters.setdefault(name, list()).append(
+            Setter(name, value, is_optional, conditions, self.is_user)
+        )
 
     def add_conditional_scope(self, scope, conditions):
         # TODO(cmaloney): 'defaults' are the same as 'can' and 'must' is identical to 'arguments' except
         # that one takes functions and one takes strings. Simplify to just 'can', 'must'.
-        assert scope.keys() <= {'validate', 'default', 'must', 'conditional'}
+        assert scope.keys() <= {'validate', 'default', 'must', 'secret', 'conditional'}
```
Comment (orsenthil) [inline gen/internals.py]: `.issubset` would be more intuitive (IMO).

File: dcos_installer/backend.py [hunk 3]
```diff
@@ -207,10 +207,12 @@ def calculate_aws_template_storage_region_name(
         'num_masters': '5',
         'aws_template_upload': 'true',
         'aws_template_storage_bucket_path_autocreate': 'true',
-        'bootstrap_id': lambda: gen.calc.calculate_environment_variable('BOOTSTRAP_ID')
+        'bootstrap_id': lambda: gen.calc.calculate_environment_variable('BOOTSTRAP_ID'),
         # TODO(cmaloney): Add defaults for getting AWS_ACCESS_KEY_ID, AWS_SECRET_ACCESS_KEY from the
```
[the remark this answers is not in the document]
Comment (branden) [reply dcos_installer/backend.py]: We're defaulting to empty strings, not pulling values from the environment, so this TODO is still unaddressed.

File: gen/internals.py (hunk 2 above)
Comment (branden) [reply gen/internals.py]: Eh this syntax doesn't really bug me. And I'm just trying to add a key to the allowed set here.

File: dcos_installer/backend.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (orsenthil) [reply dcos_installer/backend.py]: Okay, I am not sure if we addressing that `TODO` for the concern that is mentioned in the statement. `Not doing for now..`

File: gen/internals.py (hunk 2 above)
Comment (orsenthil) [reply gen/internals.py]: The review comment was not for your change. :) It was the existing use of tricky syntax.

File: gen/__init__.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (branden) [reply gen/__init__.py]: I just recalled why that didn't work -- each `source.secret` is a list. I'm not trying to create a set of those lists, I'm trying to create a set of the elements within them. So I'm passing a sequence of `source.secret`s to `set().union()`, which returns the set I need.

[the remark this answers is not in the document]
Comment (branden) [reply gen/__init__.py]: I'll simplify it to `list(set(var_name for source in sources for var_name in source.secret))`

**Comment by orsenthil** [comment]
@mesosphere-mergebot bump-ee

**Comment by branden** [comment]
@mesosphere-mergebot label shipit

**Pull request merged**